internal/error.go

```diff
@@ -273,6 +277,29 @@ var (
 	ErrActivityResultPending = errors.New("not error: do not autocomplete, using Client.CompleteActivity() to complete")
 )
 
+func bypassDurability(err error) error {
+	if err == nil {
+		return nil
+	}
+	return durabilityBypassError{error: err}
+}
+
+func (e durabilityBypassError) Unwrap() error {
+	return e.error
+}
+
+func (e durabilityBypassError) PreferBypassDurability() bool {
+	return true
+}
+
+func extractDurabilityBypassPreference(err error) enumspb.UpdateWorkflowRejectionDurabilityPreference {
+	var intf interface{ PreferBypassDurability() bool }
+	if errors.As(err, &intf) && intf.PreferBypassDurability() {
```

Comment:
I don't think you should use `errors.As` because that walks the chain, and you don't care what weird thing they did 10 errors deep. The previously authors of the Go SDK made this mistake far too many times and it has cost us with confusion when checking error types. A simple `bypassErr, _ := err.(interface { PreferBypassDurability() bool }); bypassErr != nil && bypassErr.PreferBypassDurability()` I think would work and be clear that that has to be the actual returned error to not persist.

Reply:
I'm roughly ok with that. Just means that the top-level error returned MUST implement that interface if the workflow author wants bypasss. One last fmt.Errorf with %w would hide the intent. Do you foresee a situation where the bypassError is somewhere in the error cause chain but we _wouldn't_ want to honor it?

Reply:
While I don't foresee that situation, I think clearly requiring that the error is that type instead of allowing it anywhere in the chain is much easier to comprehend by users. If they wrap it, they lost it. Same way https://github.com/grpc/grpc-go/blob/v1.48.0/status/status.go#L91 works for gRPC instead of walking the chain (though they have more reason to suspect people might unintentionally wrap than we do).

It almost makes me want to use a custom result type as I mentioned in another comment. When adding what is essentially a tertiary return type such as we're doing here, that's hard to model by type of error.

Reply:
Current direction is to move away from distinguishing between bypass and non-bypass errors. Any error coming from a validator will be implicitly prefer-bypass; errors from execution function remain implicitly non-bypass.